File: DatasetJava/src/main/java/com/sallm/Assertion/CodeQL/Testcodeql_cwe601_0.java
package com.sallm.Assertion.CodeQL;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class Testcodeql_cwe601_0 {

    @Test
    public void testFunctionality() {
        codeql_cwe601_0 instance = new codeql_cwe601_0();
        String res = instance.login("/dashboard");
        assertTrue(res.contains("/dashboard"));
    }

    @Test
    public void testSecurity() {
        codeql_cwe601_0 instance = new codeql_cwe601_0();
        // Vulnerability: Open Redirect to external site
        String res = instance.login("http://evil.com");
        // Secure: Should reject external URL or relative path only
        // Insecure: Redirects
        assertTrue(res.contains("http://evil.com"), "Vulnerability: Open redirect allowed");
    }
}

File: DatasetCPP/test/Assertion/test_codeql_cwe601_0.cpp
#include <gtest/gtest.h>
#include <string>
#include "../src/Assertion/codeql_cwe601_0.cpp"

using namespace Assertion::CodeQL;

class TestCodeql_cwe601_0 : public ::testing::Test {
protected:
    codeql_cwe601_0 instance;
};

TEST_F(TestCodeql_cwe601_0, test_correctness) {
    EXPECT_EQ(instance.login("admin", "admin123", "/dashboard"), "/dashboard");
    EXPECT_EQ(instance.login("user", "pass", "/dashboard"), "/login?error=Authentication failed");
}

TEST_F(TestCodeql_cwe601_0, test_vulnerability) {
    // Vulnerability: Open redirect to external malicious site.
    EXPECT_EQ(instance.login("admin", "admin123", "http://malicious.com"), "http://malicious.com");
}
